src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,5 +353,113 @@ private BoundExpression RewriteStringConcatInExpressionLambda(CSharpSyntaxNode s
+                        // other types supported by ConcatExprCanBeOptimizedWithToString,
+                        // but, in practice, of that list Char is the only one that matters.)
+                        if (operand.Kind == BoundKind.Literal)
+                        {
+                            ConstantValue cv = ((BoundLiteral)operand).ConstantValue;
+                            if (cv != null && cv.SpecialType == SpecialType.System_Char)
+                            {
+                                return _factory.StringLiteral(cv.CharValue.ToString());
+                            }
+                        }
+
+                        // Can the expression be optimized with a ToString call?
+                        // If so, we can synthesize a ToString call to avoid boxing.
+                        if (ConcatExprCanBeOptimizedWithToString(operand.Type))
+                        {
+                            var toString = GetSpecialTypeMethod(syntax, SpecialMember.System_Object__ToString);
+                            return BoundCall.Synthesized(syntax, operand, toString);
+                        }
+                    }
+                }
+            }
+
+            // Optimization not possible; just return the original expression.
+            return expr;
+        }
+
+        /// <summary>
+        /// Gets whether the type of an argument used in string concatenation can
+        /// bee optimized by first calling ToString on it before passing the argument
+        /// to the String.Concat function.
+        /// </summary>
+        /// <param name="symbol">The type symbol of the argument.</param>
+        /// <returns>
+        /// true if ToString may be used; false if using ToString could lead to observable differences in behavior.
+        /// </returns>
+        private static bool ConcatExprCanBeOptimizedWithToString(TypeSymbol symbol)
+        {
+            // There are several constraints we need to be aware of:
+            // - ToString has the most benefit when used on value types, to avoid boxing
+            //   when passing the value type instance as an Object to String.Concat(Object, ...)
+            //   Therefore, we're primary interested in value types.
+            // - It's possible for both value types and reference types to have a mutating
+            //   ToString method, and as such it would be a behavior-breaking change to
+            //   call ToString where/when we weren't before.  On a reference type, this could
+            //   change the order of operations as to when ToString is called (before or after
+            //   calling ToString on subsequent arguments).  For value types, it could mean mutating
+            //   the original instead of the boxed copy.  Therefore, we can only apply the
+            //   optimization in cases where we know ToString to be non-mutating.
+            // - It's possible that subsequent expressions in the concatenation mutate state
```

Comment:
:bulb: Since the current behavior is inconsistent with the C# spec, you should include a note here that the attention to preserving this behavior is for strict preservation of the backwards-compatible behavior. Also, I still think we should wait for the "compatibility committee" (or whoever) to review the situation and determine whether or not we should change the Roslyn behavior to actually follow the spec.

:question: How long will it take for the compatibility evaluation to be complete?